# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/CanonicalBrowsingContext.cpp
# Commit: afd6a3f623e1
# Full Hash: afd6a3f623e1b0d9de616c1dfa6f2806e4a587af
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-07-31 04:11:13
# Regressor Bug: 1652144
# File Overlap Count: 1
# Description:
#   Bug 1652144 - Part 2: Consider current BrowsingContextGroup for process selection, r=farre
#   
#   This requires keeping track of the current process used to host documents with a
#   particular remote type loaded in each BrowsingContextGroup. Due to lifecycle
#   oddities, this set is kept separate from the existing subscribers set on
# ==============================================================================

diff -r 3d58dcc5e59f -r afd6a3f623e1 docshell/base/CanonicalBrowsingContext.cpp
--- a/docshell/base/CanonicalBrowsingContext.cpp	Thu Jul 30 20:27:34 2020 +0000
+++ b/docshell/base/CanonicalBrowsingContext.cpp	Thu Jul 30 20:27:37 2020 +0000
@@ -962,8 +962,26 @@
   if (aRemoteType.IsEmpty()) {
     change->ProcessReady();
   } else {
+    // Try to predict which BrowsingContextGroup will be used for the final load
+    // in this BrowsingContext. This has to be accurate if switching into an
+    // existing group, as it will control what pool of processes will be used
+    // for process selection.
+    //
+    // It's _technically_ OK to provide a group here if we're actually going to
+    // switch into a brand new group, though it's sub-optimal, as it can
+    // restrict the set of processes we're using.
+    RefPtr<BrowsingContextGroup> finalGroup;
+    if (aSpecificGroupId) {
+      // If we have a specific group, we're going to end up loading in it.
+      finalGroup = BrowsingContextGroup::GetOrCreate(aSpecificGroupId);
+    } else if (!aReplaceBrowsingContext) {
+      // If we're not replacing, we'll end up back in this group.
+      finalGroup = Group();
+    }
+
     change->mContentParent = ContentParent::GetNewOrUsedLaunchingBrowserProcess(
         /* aRemoteType = */ aRemoteType,
+        /* aGroup = */ finalGroup,
         /* aPriority = */ hal::PROCESS_PRIORITY_FOREGROUND,
         /* aPreferUsed = */ false);
     if (!change->mContentParent) {